documentation/guide-api-first.asciidoc

```diff
@@ -0,0 +1,144 @@
+
+After careful planning, we defined our future API. Now we would like to implement a provider of this API.
+We could manually create all the link:guide-rest.asciidoc#jax-rs[JAX-RS] endpoints based on the schema and then test whether the actually provided interface conforms with the API schema. However, it is an error prone and laborious process, especially if the API gets big.
+Luckily, we can use great open source tooling to generate JAX-RS interfaces that conform to 100% with the schema, and then we simply implement them.
+
+There are many open source tools, that allow code generation from OpenAPI schema files, for example https://openapi-generator.tech/[OpenAPI Generator] or https://github.com/swagger-api/swagger-codegen[Swagger codegen]. Both are based on Java and even provide maven plugins to integrate them in our build.
+
+=== OpenAPI Generator maven plugin
+
+If we already have our backend maven project, using the maven plugin will be easiest for us. Lets define a new maven profile and add the OpenAPI generator plugin:
+[source, xml]
+----
+<profile>
+  <id>apigen</id>
+  <activation><activeByDefault>true</activeByDefault></activation>
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>org.openapitools</groupId>
+        <artifactId>openapi-generator-maven-plugin</artifactId>
+        <version>5.1.0</version>
+        <executions>
+          <execution>
+            <goals>
+              <goal>generate</goal>
+            </goals>
+            <configuration>
+              <!-- input spec can be a url, or a local file -->
+              <inputSpec>https://raw.githubusercontent.com/OAI/OpenAPI-Specification/main/examples/v3.0/petstore.yaml</inputSpec>
+              <!-- <inputSpec>${project.basedir}/src/main/resources/api.yaml</inputSpec> -->
+              <generatorName>jaxrs-spec</generatorName>
+              <apiPackage>com.petclinic.api</apiPackage>
+              <modelPackage>com.petclinic.api.model</modelPackage>
+              <!-- <library>quarkus</library> -->
+              <configOptions>
+                <sourceFolder>src/gen/java/main</sourceFolder>
+                <useSwaggerAnnotations>false</useSwaggerAnnotations>
+                <useTags>true</useTags>
+                <interfaceOnly>true</interfaceOnly>
+                <generatePom>false</generatePom>
+              </configOptions>
+            </configuration>
+          </execution>
+        </executions>
+      </plugin>
+    </plugins>
+  </build>
+</profile>
+----
+
+OpenAPI generator project includes support for many languages and frameworks - in our case we picked `jaxrs-spec` server generator.
+Many of the generator templates have further sub-generators, for example for your specific framework. We could, for example use `quarkus` sub-generator to bootstrap a full working Quarkus maven project, but in our case, we want to have more control over the process, therefore we opt for standard `jaxrs` template and pass some additional configuration: 
```

Comment:
Many of the generator templates have further sub-generators**, such as one** for your specific framework. We could, for (instance OR example), use `quarkus` sub-generator to bootstrap a **fully** working Quarkus maven project...

https://linguaholic.com/linguablog/comma-before-for-example/
https://linguaholic.com/linguablog/comma-after-for-example/
;)